<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Cell Type Prediction &#8212; spatialproteomics</title>
    
  <link rel="stylesheet" href="../_static/css/index.f658d18f9b420779cfdf24aa0a7e2d77.css">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="../_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="../_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-book-theme.40e2e510f6b7d1648584402491bb10fe.css?v=a123c646" />
    
  <link rel="preload" as="script" href="../_static/js/index.d3f166471bb80abb5163.js">

    <script src="../_static/documentation_options.js?v=1027fec7"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/sphinx-book-theme.d31b09fe5c1d09cb49b26a786de4a05d.js?v=dbd2e957"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Neighborhood Analysis" href="Neighborhoods.html" />
    <link rel="prev" title="Image Processing" href="ImageProcessing.html" />
    <link href="../_static/style.css" rel="stylesheet" type="text/css">

  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../index.html">
  
  
  <h1 class="site-logo" id="site-title">spatialproteomics</h1>
  
</a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form>
<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <p class="caption collapsible-parent" role="heading">
 <span class="caption-text">
  Contents:
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing.html">
   The preprocessing (
   <code class="code docutils literal notranslate">
    <span class="pre">
     pp
    </span>
   </code>
   ) accessor
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../label.html">
   The label (
   <code class="code docutils literal notranslate">
    <span class="pre">
     la
    </span>
   </code>
   ) accessor
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../neighborhood.html">
   The neighborhood (
   <code class="code docutils literal notranslate">
    <span class="pre">
     nh
    </span>
   </code>
   ) accessor
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../plot.html">
   The plot (
   <code class="code docutils literal notranslate">
    <span class="pre">
     pl
    </span>
   </code>
   ) accessor
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../tool.html">
   The tool (
   <code class="code docutils literal notranslate">
    <span class="pre">
     tl
    </span>
   </code>
   ) accessor
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../image_container.html">
   The image container
  </a>
 </li>
</ul>
<p class="caption collapsible-parent" role="heading">
 <span class="caption-text">
  Tutorials:
 </span>
</p>
<ul class="current nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="Installation.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="ExampleWorkflow.html">
   Example Workflow with spatialproteomics
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="ExampleWorkflowSpatialdata.html">
   Example Workflow with spatialproteomics and spatialdata
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Slicing.html">
   Subselecting Data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Segmentation.html">
   Segmentation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="ArtifactRemoval.html">
   Interactive Removal of Artifacts
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Plotting.html">
   Plotting
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="ImageProcessing.html">
   Image Processing
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Cell Type Prediction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Neighborhoods.html">
   Neighborhood Analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Interoperability.html">
   Interoperability
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Interactivity.html">
   Interactivity
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="FAQ.html">
   FAQ
  </a>
 </li>
</ul>

</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Theme by the <a href="https://ebp.jupyterbook.org">Executable Book Project</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/notebooks/CellTypePrediction.ipynb.txt"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

        <!-- Source interaction buttons -->


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

    </div>

    <!-- Table of contents -->
    <div class="d-none d-md-block col-md-2 bd-toc show">
        
        <div class="tocsection onthispage pt-5 pb-3">
            <i class="fas fa-list"></i> Contents
        </div>
        <nav id="bd-toc-nav">
            <ul class="nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#Image-Processing">
   Image Processing
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#Cell-Type-Prediction-with-Argmax">
   Cell Type Prediction with Argmax
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#Cell-Type-Prediction-with-Astir">
   Cell Type Prediction with Astir
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#Marker-Binarization">
   Marker Binarization
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#Cell-Subtype-Prediction">
   Cell Subtype Prediction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#Advanced-Cell-Type-Gating">
   Advanced Cell Type Gating
  </a>
 </li>
</ul>

        </nav>
        
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Cell-Type-Prediction">
<h1>Cell Type Prediction<a class="headerlink" href="#Cell-Type-Prediction" title="Link to this heading">¶</a></h1>
<p>The goal of cell type prediction is to assign a label to each cell. While methods such as clustering are frequently used in single cell transcriptomics data, these methods do not necessarily translate well to highly multiplexed fluorescence images. Instead, this notebook follows a two-step approach. In the first step, major cell types (such as B cells, T cells, macrophages, etc.) are predicted by using a set of mutually exclusive markers. These initial labels are predicted with methods such as
<code class="docutils literal notranslate"><span class="pre">argmax</span></code> or <code class="docutils literal notranslate"><span class="pre">astir</span></code> and build the foundation for step two, in which we will look at functional markers. Here, we use a binarization-like approach. This means that for our functional markers, each cell gets either a value of 1 (if the cell expresses that marker) or 0 (if the cell does not express that marker). While this is a simplification of the underlying biology, this abstraction can be useful to counteract the often poor signal-to-noise ratio inherent in the data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">reload_ext</span> autoreload
<span class="o">%</span><span class="k">autoreload</span> 2
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">spatialproteomics</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sp</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">xarray</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">xr</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.signal</span><span class="w"> </span><span class="kn">import</span> <span class="n">medfilt2d</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ds</span> <span class="o">=</span> <span class="n">xr</span><span class="o">.</span><span class="n">open_zarr</span><span class="p">(</span><span class="s2">&quot;../../data/BNHL_166_4_I2_LK_2.zarr&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<section id="Image-Processing">
<h2>Image Processing<a class="headerlink" href="#Image-Processing" title="Link to this heading">¶</a></h2>
<p>Before predicting cell types, it is typically helpful to perform some preprocessing on the images. Here, we simply perform some custom thresholding on the markers to clean up the images and boost the signal-to-noise ratio.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># preprocessing: thresholding by percentiles and applying a median filter</span>
<span class="n">channels</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;PAX5&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD3&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD11b&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD11c&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD15&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD68&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Podoplanin&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD31&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD34&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD90&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD56&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD4&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD8&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Ki-67&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD45RA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD45RO&quot;</span><span class="p">,</span>
<span class="p">]</span>
<span class="n">quantiles</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.95</span><span class="p">,</span> <span class="mf">0.95</span><span class="p">,</span> <span class="mf">0.95</span><span class="p">,</span> <span class="mf">0.95</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.95</span><span class="p">,</span> <span class="mf">0.95</span><span class="p">,</span> <span class="mf">0.95</span><span class="p">,</span> <span class="mf">0.95</span><span class="p">,</span> <span class="mf">0.95</span><span class="p">]</span>
<span class="n">colors</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;#e6194B&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#3cb44b&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#ffe119&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#4363d8&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#ffd8b1&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#f58231&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#911eb4&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#fffac8&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#469990&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#fabed4&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#9A6324&quot;</span><span class="p">,</span>
<span class="p">]</span>

<span class="n">channels_celltypes</span> <span class="o">=</span> <span class="n">channels</span><span class="p">[:</span><span class="o">-</span><span class="mi">5</span><span class="p">]</span>

<span class="c1"># processing and recomputing quantification</span>
<span class="n">ds_processed</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ds</span><span class="o">.</span><span class="n">pp</span><span class="p">[</span><span class="n">channels</span><span class="p">]</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">threshold</span><span class="p">(</span><span class="n">quantiles</span><span class="p">)</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">add_quantification</span><span class="p">()</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">transform_expression_matrix</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;arcsinh&quot;</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plotting the ds and ds processed next to one another</span>

<span class="c1"># visualizing the raw vs the processed image</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">pp</span><span class="p">[</span><span class="n">channels_celltypes</span><span class="p">]</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">colorize</span><span class="p">(</span><span class="n">colors</span><span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">ds_processed</span><span class="o">.</span><span class="n">pp</span><span class="p">[</span><span class="n">channels_celltypes</span><span class="p">]</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">colorize</span><span class="p">(</span><span class="n">colors</span><span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

<span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Raw&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Processed&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Text(0.5, 1.0, &#39;Processed&#39;)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_CellTypePrediction_6_1.png" src="../_images/notebooks_CellTypePrediction_6_1.png" />
</div>
</div>
</section>
<section id="Cell-Type-Prediction-with-Argmax">
<h2>Cell Type Prediction with Argmax<a class="headerlink" href="#Cell-Type-Prediction-with-Argmax" title="Link to this heading">¶</a></h2>
<p>Given the thresholded image, the simplest way to derive cell type labels is to define associations between cell types and markers, and then assign the cell type whose marker is most highly expressed (i. e. taking the argmax). This technique, also termed <code class="docutils literal notranslate"><span class="pre">argmax</span></code>, is demonstrated below.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">marker_ct_dict</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;PAX5&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD3&quot;</span><span class="p">:</span> <span class="s2">&quot;T&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD11b&quot;</span><span class="p">:</span> <span class="s2">&quot;Myeloid&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD11c&quot;</span><span class="p">:</span> <span class="s2">&quot;Dendritic&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD15&quot;</span><span class="p">:</span> <span class="s2">&quot;Granulo&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD68&quot;</span><span class="p">:</span> <span class="s2">&quot;Macro&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Podoplanin&quot;</span><span class="p">:</span> <span class="s2">&quot;Stromal PDPN&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD31&quot;</span><span class="p">:</span> <span class="s2">&quot;Stromal CD31&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD34&quot;</span><span class="p">:</span> <span class="s2">&quot;Stromal CD34&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD90&quot;</span><span class="p">:</span> <span class="s2">&quot;Stromal CD90&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD56&quot;</span><span class="p">:</span> <span class="s2">&quot;NK&quot;</span><span class="p">,</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># predicting the cell type using the argmay</span>
<span class="n">ds_with_ct_predictions</span> <span class="o">=</span> <span class="n">ds_processed</span><span class="o">.</span><span class="n">la</span><span class="o">.</span><span class="n">predict_cell_types_argmax</span><span class="p">(</span>
    <span class="n">marker_ct_dict</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="s2">&quot;_intensity&quot;</span><span class="p">,</span> <span class="n">overwrite_existing_labels</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># adding colors to match the markers</span>

<span class="n">ds_with_ct_predictions</span> <span class="o">=</span> <span class="n">ds_with_ct_predictions</span><span class="o">.</span><span class="n">la</span><span class="o">.</span><span class="n">set_label_colors</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">marker_ct_dict</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="n">colors</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Removing labels from observations. If you want to keep the labels in the obs layer, set drop_obs=False.
Label Stromal PDPN not found in the data object. Skipping.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plotting the ct predictions next to the processed image</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">ds_with_ct_predictions</span><span class="o">.</span><span class="n">pp</span><span class="p">[</span><span class="n">channels_celltypes</span><span class="p">]</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">colorize</span><span class="p">(</span><span class="n">colors</span><span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">ds_with_ct_predictions</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">render_image</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">render_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

<span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Markers&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Cell Type Predictions&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Text(0.5, 1.0, &#39;Cell Type Predictions&#39;)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_CellTypePrediction_10_1.png" src="../_images/notebooks_CellTypePrediction_10_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># this is how the image in the README was created</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">21</span><span class="p">,</span> <span class="mi">7</span><span class="p">))</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">ds_processed</span><span class="o">.</span><span class="n">pp</span><span class="p">[</span><span class="n">channels_celltypes</span><span class="p">]</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">colorize</span><span class="p">(</span><span class="n">colors</span><span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">autocrop</span><span class="p">(</span><span class="n">padding</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">pp</span><span class="p">[</span><span class="s2">&quot;DAPI&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">autocrop</span><span class="p">(</span><span class="n">padding</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">render_segmentation</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">ds_with_ct_predictions</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">autocrop</span><span class="p">(</span><span class="n">padding</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">render_image</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">render_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>


<span class="c1"># removing the x and y ticks</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>

<span class="c1"># setting titles</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Markers&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Segmentation&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Cell Types&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_CellTypePrediction_11_0.png" src="../_images/notebooks_CellTypePrediction_11_0.png" />
</div>
</div>
<p>Getting observations from the xarray object is very easy. We can simply use the method <code class="docutils literal notranslate"><span class="pre">pp.get_layer_as_df()</span></code>, which can also parse our cell types into human-readable format automatically.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># getting the obs with get_layer_as_df() to get human readable cell type annotations in a pandas data frame</span>
<span class="n">ds_with_ct_predictions</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">get_layer_as_df</span><span class="p">(</span><span class="s2">&quot;_obs&quot;</span><span class="p">,</span> <span class="n">celltypes_to_str</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>_labels</th>
      <th>centroid-0</th>
      <th>centroid-1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>T</td>
      <td>613.329787</td>
      <td>768.420213</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Macro</td>
      <td>769.098446</td>
      <td>707.544041</td>
    </tr>
    <tr>
      <th>3</th>
      <td>T</td>
      <td>774.528409</td>
      <td>644.284091</td>
    </tr>
    <tr>
      <th>4</th>
      <td>T</td>
      <td>775.902878</td>
      <td>592.744604</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Myeloid</td>
      <td>668.844749</td>
      <td>729.310502</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>12556</th>
      <td>B</td>
      <td>2340.362319</td>
      <td>1846.492754</td>
    </tr>
    <tr>
      <th>12557</th>
      <td>Macro</td>
      <td>2216.324138</td>
      <td>2296.089655</td>
    </tr>
    <tr>
      <th>12558</th>
      <td>T</td>
      <td>2265.910853</td>
      <td>2231.992248</td>
    </tr>
    <tr>
      <th>12559</th>
      <td>B</td>
      <td>2281.480687</td>
      <td>2218.424893</td>
    </tr>
    <tr>
      <th>12560</th>
      <td>B</td>
      <td>2249.069444</td>
      <td>2236.564815</td>
    </tr>
  </tbody>
</table>
<p>12560 rows × 3 columns</p>
</div></div>
</div>
</section>
<section id="Cell-Type-Prediction-with-Astir">
<h2>Cell Type Prediction with Astir<a class="headerlink" href="#Cell-Type-Prediction-with-Astir" title="Link to this heading">¶</a></h2>
<p>There also exist more involved methods of predicting cell types, such as astir. Here is a quick example on how one could apply astir to the previous image.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ds_processed</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div><svg style="position: absolute; width: 0; height: 0; overflow: hidden">
<defs>
<symbol id="icon-database" viewBox="0 0 32 32">
<path d="M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z"></path>
<path d="M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
<path d="M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
</symbol>
<symbol id="icon-file-text2" viewBox="0 0 32 32">
<path d="M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z"></path>
<path d="M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
</symbol>
</defs>
</svg>
<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.
 *
 */

:root {
  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));
  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));
  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));
  --xr-border-color: var(--jp-border-color2, #e0e0e0);
  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);
  --xr-background-color: var(--jp-layout-color0, white);
  --xr-background-color-row-even: var(--jp-layout-color1, white);
  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);
}

html[theme=dark],
body[data-theme=dark],
body.vscode-dark {
  --xr-font-color0: rgba(255, 255, 255, 1);
  --xr-font-color2: rgba(255, 255, 255, 0.54);
  --xr-font-color3: rgba(255, 255, 255, 0.38);
  --xr-border-color: #1F1F1F;
  --xr-disabled-color: #515151;
  --xr-background-color: #111111;
  --xr-background-color-row-even: #111111;
  --xr-background-color-row-odd: #313131;
}

.xr-wrap {
  display: block !important;
  min-width: 300px;
  max-width: 700px;
}

.xr-text-repr-fallback {
  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */
  display: none;
}

.xr-header {
  padding-top: 6px;
  padding-bottom: 6px;
  margin-bottom: 4px;
  border-bottom: solid 1px var(--xr-border-color);
}

.xr-header > div,
.xr-header > ul {
  display: inline;
  margin-top: 0;
  margin-bottom: 0;
}

.xr-obj-type,
.xr-array-name {
  margin-left: 2px;
  margin-right: 10px;
}

.xr-obj-type {
  color: var(--xr-font-color2);
}

.xr-sections {
  padding-left: 0 !important;
  display: grid;
  grid-template-columns: 150px auto auto 1fr 20px 20px;
}

.xr-section-item {
  display: contents;
}

.xr-section-item input {
  display: none;
}

.xr-section-item input + label {
  color: var(--xr-disabled-color);
}

.xr-section-item input:enabled + label {
  cursor: pointer;
  color: var(--xr-font-color2);
}

.xr-section-item input:enabled + label:hover {
  color: var(--xr-font-color0);
}

.xr-section-summary {
  grid-column: 1;
  color: var(--xr-font-color2);
  font-weight: 500;
}

.xr-section-summary > span {
  display: inline-block;
  padding-left: 0.5em;
}

.xr-section-summary-in:disabled + label {
  color: var(--xr-font-color2);
}

.xr-section-summary-in + label:before {
  display: inline-block;
  content: '►';
  font-size: 11px;
  width: 15px;
  text-align: center;
}

.xr-section-summary-in:disabled + label:before {
  color: var(--xr-disabled-color);
}

.xr-section-summary-in:checked + label:before {
  content: '▼';
}

.xr-section-summary-in:checked + label > span {
  display: none;
}

.xr-section-summary,
.xr-section-inline-details {
  padding-top: 4px;
  padding-bottom: 4px;
}

.xr-section-inline-details {
  grid-column: 2 / -1;
}

.xr-section-details {
  display: none;
  grid-column: 1 / -1;
  margin-bottom: 5px;
}

.xr-section-summary-in:checked ~ .xr-section-details {
  display: contents;
}

.xr-array-wrap {
  grid-column: 1 / -1;
  display: grid;
  grid-template-columns: 20px auto;
}

.xr-array-wrap > label {
  grid-column: 1;
  vertical-align: top;
}

.xr-preview {
  color: var(--xr-font-color3);
}

.xr-array-preview,
.xr-array-data {
  padding: 0 5px !important;
  grid-column: 2;
}

.xr-array-data,
.xr-array-in:checked ~ .xr-array-preview {
  display: none;
}

.xr-array-in:checked ~ .xr-array-data,
.xr-array-preview {
  display: inline-block;
}

.xr-dim-list {
  display: inline-block !important;
  list-style: none;
  padding: 0 !important;
  margin: 0;
}

.xr-dim-list li {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.xr-dim-list:before {
  content: '(';
}

.xr-dim-list:after {
  content: ')';
}

.xr-dim-list li:not(:last-child):after {
  content: ',';
  padding-right: 5px;
}

.xr-has-index {
  font-weight: bold;
}

.xr-var-list,
.xr-var-item {
  display: contents;
}

.xr-var-item > div,
.xr-var-item label,
.xr-var-item > .xr-var-name span {
  background-color: var(--xr-background-color-row-even);
  margin-bottom: 0;
}

.xr-var-item > .xr-var-name:hover span {
  padding-right: 5px;
}

.xr-var-list > li:nth-child(odd) > div,
.xr-var-list > li:nth-child(odd) > label,
.xr-var-list > li:nth-child(odd) > .xr-var-name span {
  background-color: var(--xr-background-color-row-odd);
}

.xr-var-name {
  grid-column: 1;
}

.xr-var-dims {
  grid-column: 2;
}

.xr-var-dtype {
  grid-column: 3;
  text-align: right;
  color: var(--xr-font-color2);
}

.xr-var-preview {
  grid-column: 4;
}

.xr-index-preview {
  grid-column: 2 / 5;
  color: var(--xr-font-color2);
}

.xr-var-name,
.xr-var-dims,
.xr-var-dtype,
.xr-preview,
.xr-attrs dt {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding-right: 10px;
}

.xr-var-name:hover,
.xr-var-dims:hover,
.xr-var-dtype:hover,
.xr-attrs dt:hover {
  overflow: visible;
  width: auto;
  z-index: 1;
}

.xr-var-attrs,
.xr-var-data,
.xr-index-data {
  display: none;
  background-color: var(--xr-background-color) !important;
  padding-bottom: 5px !important;
}

.xr-var-attrs-in:checked ~ .xr-var-attrs,
.xr-var-data-in:checked ~ .xr-var-data,
.xr-index-data-in:checked ~ .xr-index-data {
  display: block;
}

.xr-var-data > table {
  float: right;
}

.xr-var-name span,
.xr-var-data,
.xr-index-name div,
.xr-index-data,
.xr-attrs {
  padding-left: 25px !important;
}

.xr-attrs,
.xr-var-attrs,
.xr-var-data,
.xr-index-data {
  grid-column: 1 / -1;
}

dl.xr-attrs {
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: 125px auto;
}

.xr-attrs dt,
.xr-attrs dd {
  padding: 0;
  margin: 0;
  float: left;
  padding-right: 10px;
  width: auto;
}

.xr-attrs dt {
  font-weight: normal;
  grid-column: 1;
}

.xr-attrs dt:hover span {
  display: inline-block;
  background: var(--xr-background-color);
  padding-right: 10px;
}

.xr-attrs dd {
  grid-column: 2;
  white-space: pre-wrap;
  word-break: break-all;
}

.xr-icon-database,
.xr-icon-file-text2,
.xr-no-icon {
  display: inline-block;
  vertical-align: middle;
  width: 1em;
  height: 1.5em !important;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
}
</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;
Dimensions:         (labels: 8, la_props: 2, cells: 12560, features: 3,
                     channels: 16, x: 3000, y: 3000)
Coordinates:
  * cells           (cells) int64 1 2 3 4 5 6 ... 12556 12557 12558 12559 12560
  * channels        (channels) &lt;U11 &#x27;PAX5&#x27; &#x27;CD3&#x27; &#x27;CD11b&#x27; ... &#x27;CD45RA&#x27; &#x27;CD45RO&#x27;
  * features        (features) &lt;U10 &#x27;_labels&#x27; &#x27;centroid-0&#x27; &#x27;centroid-1&#x27;
  * la_props        (la_props) &lt;U6 &#x27;_color&#x27; &#x27;_name&#x27;
  * labels          (labels) int64 1 2 3 4 5 6 7 8
  * x               (x) int64 0 1 2 3 4 5 6 ... 2994 2995 2996 2997 2998 2999
  * y               (y) int64 0 1 2 3 4 5 6 ... 2994 2995 2996 2997 2998 2999
Data variables:
    _la_properties  (labels, la_props) &lt;U20 dask.array&lt;chunksize=(8, 2), meta=np.ndarray&gt;
    _obs            (cells, features) float64 dask.array&lt;chunksize=(6280, 3), meta=np.ndarray&gt;
    _segmentation   (y, x) int64 0 0 0 0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0 0 0 0
    _image          (channels, y, x) uint8 0 0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0
    _intensity      (cells, channels) float64 0.004255 2.46 ... 1.384 0.06662</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-ea0517d9-b92b-4ea0-b5ac-486737004da9' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-ea0517d9-b92b-4ea0-b5ac-486737004da9' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span class='xr-has-index'>labels</span>: 8</li><li><span class='xr-has-index'>la_props</span>: 2</li><li><span class='xr-has-index'>cells</span>: 12560</li><li><span class='xr-has-index'>features</span>: 3</li><li><span class='xr-has-index'>channels</span>: 16</li><li><span class='xr-has-index'>x</span>: 3000</li><li><span class='xr-has-index'>y</span>: 3000</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-18373397-bdf6-471a-a57b-2c7c336ff267' class='xr-section-summary-in' type='checkbox'  checked><label for='section-18373397-bdf6-471a-a57b-2c7c336ff267' class='xr-section-summary' >Coordinates: <span>(7)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>cells</span></div><div class='xr-var-dims'>(cells)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>1 2 3 4 ... 12557 12558 12559 12560</div><input id='attrs-16a08d48-8a26-4829-ba28-13791b33ed96' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-16a08d48-8a26-4829-ba28-13791b33ed96' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0be62b14-64bb-4ce1-a175-e6d6bc68297a' class='xr-var-data-in' type='checkbox'><label for='data-0be62b14-64bb-4ce1-a175-e6d6bc68297a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([    1,     2,     3, ..., 12558, 12559, 12560])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>channels</span></div><div class='xr-var-dims'>(channels)</div><div class='xr-var-dtype'>&lt;U11</div><div class='xr-var-preview xr-preview'>&#x27;PAX5&#x27; &#x27;CD3&#x27; ... &#x27;CD45RA&#x27; &#x27;CD45RO&#x27;</div><input id='attrs-08e63212-f92d-406d-a3cc-9a7a3b3192e1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-08e63212-f92d-406d-a3cc-9a7a3b3192e1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-bd370f62-01f9-4e4e-b756-ae94b21c0dac' class='xr-var-data-in' type='checkbox'><label for='data-bd370f62-01f9-4e4e-b756-ae94b21c0dac' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;PAX5&#x27;, &#x27;CD3&#x27;, &#x27;CD11b&#x27;, &#x27;CD11c&#x27;, &#x27;CD15&#x27;, &#x27;CD68&#x27;, &#x27;Podoplanin&#x27;, &#x27;CD31&#x27;,
       &#x27;CD34&#x27;, &#x27;CD90&#x27;, &#x27;CD56&#x27;, &#x27;CD4&#x27;, &#x27;CD8&#x27;, &#x27;Ki-67&#x27;, &#x27;CD45RA&#x27;, &#x27;CD45RO&#x27;],
      dtype=&#x27;&lt;U11&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>features</span></div><div class='xr-var-dims'>(features)</div><div class='xr-var-dtype'>&lt;U10</div><div class='xr-var-preview xr-preview'>&#x27;_labels&#x27; &#x27;centroid-0&#x27; &#x27;centroid-1&#x27;</div><input id='attrs-9b06893f-e85e-49f2-8d0a-a3e77c488167' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-9b06893f-e85e-49f2-8d0a-a3e77c488167' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1c0ca1ae-d68d-4b45-a4bd-90a52c9b092f' class='xr-var-data-in' type='checkbox'><label for='data-1c0ca1ae-d68d-4b45-a4bd-90a52c9b092f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;_labels&#x27;, &#x27;centroid-0&#x27;, &#x27;centroid-1&#x27;], dtype=&#x27;&lt;U10&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>la_props</span></div><div class='xr-var-dims'>(la_props)</div><div class='xr-var-dtype'>&lt;U6</div><div class='xr-var-preview xr-preview'>&#x27;_color&#x27; &#x27;_name&#x27;</div><input id='attrs-8cf59a4b-d384-4b5e-a578-10f5a6d60800' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8cf59a4b-d384-4b5e-a578-10f5a6d60800' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-740f98f0-6c17-4e7d-9f02-1dfba1c73e95' class='xr-var-data-in' type='checkbox'><label for='data-740f98f0-6c17-4e7d-9f02-1dfba1c73e95' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;_color&#x27;, &#x27;_name&#x27;], dtype=&#x27;&lt;U6&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>labels</span></div><div class='xr-var-dims'>(labels)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>1 2 3 4 5 6 7 8</div><input id='attrs-3c6519a3-f4e3-44f9-86f3-9ac7d4e2cdc7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3c6519a3-f4e3-44f9-86f3-9ac7d4e2cdc7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-43a64cf5-6492-48c3-abb1-0c8e6b0c9f18' class='xr-var-data-in' type='checkbox'><label for='data-43a64cf5-6492-48c3-abb1-0c8e6b0c9f18' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([1, 2, 3, 4, 5, 6, 7, 8])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>x</span></div><div class='xr-var-dims'>(x)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 4 ... 2996 2997 2998 2999</div><input id='attrs-65ffb291-709c-4625-9fa3-f8780b3e5113' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-65ffb291-709c-4625-9fa3-f8780b3e5113' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-4704a01d-ee48-41ed-9c1d-6175a0df4914' class='xr-var-data-in' type='checkbox'><label for='data-4704a01d-ee48-41ed-9c1d-6175a0df4914' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([   0,    1,    2, ..., 2997, 2998, 2999])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>y</span></div><div class='xr-var-dims'>(y)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 4 ... 2996 2997 2998 2999</div><input id='attrs-5931da04-22de-4231-b355-55b63bc0cdc4' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5931da04-22de-4231-b355-55b63bc0cdc4' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-454848ce-9817-47c6-90de-704f82273d36' class='xr-var-data-in' type='checkbox'><label for='data-454848ce-9817-47c6-90de-704f82273d36' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([   0,    1,    2, ..., 2997, 2998, 2999])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-c54b0bc2-2c26-4e4e-af3a-e96d6767b4c4' class='xr-section-summary-in' type='checkbox'  checked><label for='section-c54b0bc2-2c26-4e4e-af3a-e96d6767b4c4' class='xr-section-summary' >Data variables: <span>(5)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>_la_properties</span></div><div class='xr-var-dims'>(labels, la_props)</div><div class='xr-var-dtype'>&lt;U20</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(8, 2), meta=np.ndarray&gt;</div><input id='attrs-df64a071-1cd8-40a3-a693-ba9b050f981d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-df64a071-1cd8-40a3-a693-ba9b050f981d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-838934cf-8ba1-4b08-952b-62f400908487' class='xr-var-data-in' type='checkbox'><label for='data-838934cf-8ba1-4b08-952b-62f400908487' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
    <tr>
        <td>
            <table style="border-collapse: collapse;">
                <thead>
                    <tr>
                        <td> </td>
                        <th> Array </th>
                        <th> Chunk </th>
                    </tr>
                </thead>
                <tbody>

                    <tr>
                        <th> Bytes </th>
                        <td> 1.25 kiB </td>
                        <td> 1.25 kiB </td>
                    </tr>

                    <tr>
                        <th> Shape </th>
                        <td> (8, 2) </td>
                        <td> (8, 2) </td>
                    </tr>
                    <tr>
                        <th> Dask graph </th>
                        <td colspan="2"> 1 chunks in 2 graph layers </td>
                    </tr>
                    <tr>
                        <th> Data type </th>
                        <td colspan="2"> <U20 numpy.ndarray </td>
                    </tr>
                </tbody>
            </table>
        </td>
        <td>
        <svg width="92" height="170" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="42" y2="0" style="stroke-width:2" />
  <line x1="0" y1="120" x2="42" y2="120" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="120" style="stroke-width:2" />
  <line x1="42" y1="0" x2="42" y2="120" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 42.89879552186203,0.0 42.89879552186203,120.0 0.0,120.0" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="21.449398" y="140.000000" font-size="1.0rem" font-weight="100" text-anchor="middle" >2</text>
  <text x="62.898796" y="60.000000" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,62.898796,60.000000)">8</text>
</svg>
        </td>
    </tr>
</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>_obs</span></div><div class='xr-var-dims'>(cells, features)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(6280, 3), meta=np.ndarray&gt;</div><input id='attrs-c0b065fd-0518-49be-8d0c-773441abf31a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c0b065fd-0518-49be-8d0c-773441abf31a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a283212d-56b4-4d32-880e-c02031bbac8b' class='xr-var-data-in' type='checkbox'><label for='data-a283212d-56b4-4d32-880e-c02031bbac8b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
    <tr>
        <td>
            <table style="border-collapse: collapse;">
                <thead>
                    <tr>
                        <td> </td>
                        <th> Array </th>
                        <th> Chunk </th>
                    </tr>
                </thead>
                <tbody>

                    <tr>
                        <th> Bytes </th>
                        <td> 294.38 kiB </td>
                        <td> 147.19 kiB </td>
                    </tr>

                    <tr>
                        <th> Shape </th>
                        <td> (12560, 3) </td>
                        <td> (6280, 3) </td>
                    </tr>
                    <tr>
                        <th> Dask graph </th>
                        <td colspan="2"> 2 chunks in 2 graph layers </td>
                    </tr>
                    <tr>
                        <th> Data type </th>
                        <td colspan="2"> float64 numpy.ndarray </td>
                    </tr>
                </tbody>
            </table>
        </td>
        <td>
        <svg width="75" height="170" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="25" y2="0" style="stroke-width:2" />
  <line x1="0" y1="60" x2="25" y2="60" />
  <line x1="0" y1="120" x2="25" y2="120" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="120" style="stroke-width:2" />
  <line x1="25" y1="0" x2="25" y2="120" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 25.412616514582485,0.0 25.412616514582485,120.0 0.0,120.0" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="12.706308" y="140.000000" font-size="1.0rem" font-weight="100" text-anchor="middle" >3</text>
  <text x="45.412617" y="60.000000" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,45.412617,60.000000)">12560</text>
</svg>
        </td>
    </tr>
</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>_segmentation</span></div><div class='xr-var-dims'>(y, x)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0</div><input id='attrs-087fb881-c993-487b-8fb7-2056ac53cc20' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-087fb881-c993-487b-8fb7-2056ac53cc20' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d989b849-b8a6-4e6f-a210-102dcf1d3bd0' class='xr-var-data-in' type='checkbox'><label for='data-d989b849-b8a6-4e6f-a210-102dcf1d3bd0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>_image</span></div><div class='xr-var-dims'>(channels, y, x)</div><div class='xr-var-dtype'>uint8</div><div class='xr-var-preview xr-preview'>0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0</div><input id='attrs-eaaed12f-e5ca-4e12-b2dc-e76d50df48ee' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-eaaed12f-e5ca-4e12-b2dc-e76d50df48ee' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-6f8c8b4f-5686-46d0-ade7-545b8b7cdcd7' class='xr-var-data-in' type='checkbox'><label for='data-6f8c8b4f-5686-46d0-ade7-545b8b7cdcd7' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[[0, 0, 0, ..., 0, 0, 1],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
...
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]]], dtype=uint8)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>_intensity</span></div><div class='xr-var-dims'>(cells, channels)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.004255 2.46 ... 1.384 0.06662</div><input id='attrs-80f71fc1-00f6-4622-a40c-b85adacb36dd' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-80f71fc1-00f6-4622-a40c-b85adacb36dd' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-848a4834-c301-4039-b364-4964494a3b48' class='xr-var-data-in' type='checkbox'><label for='data-848a4834-c301-4039-b364-4964494a3b48' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[4.25530631e-03, 2.46034155e+00, 2.12765797e-03, ...,
        1.06382959e-03, 4.78540739e-02, 0.00000000e+00],
       [0.00000000e+00, 3.03109280e-01, 1.25937259e+00, ...,
        3.31545474e-02, 0.00000000e+00, 0.00000000e+00],
       [0.00000000e+00, 2.15362817e+00, 9.93224357e-01, ...,
        3.40908431e-03, 0.00000000e+00, 1.11464514e+00],
       ...,
       [1.06773747e-01, 1.35232153e+00, 2.79033557e-02, ...,
        0.00000000e+00, 5.15608185e-01, 6.36717718e-01],
       [1.18304095e+00, 1.13947789e+00, 9.25715906e-02, ...,
        1.71673735e-03, 8.79551538e-01, 2.57510445e-03],
       [8.97647445e-01, 6.25145117e-01, 1.66817809e-01, ...,
        0.00000000e+00, 1.38367696e+00, 6.66173825e-02]])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-f50b1f45-79c3-411d-ae10-39d032b292ed' class='xr-section-summary-in' type='checkbox'  ><label for='section-f50b1f45-79c3-411d-ae10-39d032b292ed' class='xr-section-summary' >Indexes: <span>(7)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>cells</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-e2776883-c5f0-4cde-bf89-eb376b3f97b9' class='xr-index-data-in' type='checkbox'/><label for='index-e2776883-c5f0-4cde-bf89-eb376b3f97b9' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([    1,     2,     3,     4,     5,     6,     7,     8,     9,    10,
       ...
       12551, 12552, 12553, 12554, 12555, 12556, 12557, 12558, 12559, 12560],
      dtype=&#x27;int64&#x27;, name=&#x27;cells&#x27;, length=12560))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>channels</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-0a073776-afed-4eb6-9570-f6238bbf2802' class='xr-index-data-in' type='checkbox'/><label for='index-0a073776-afed-4eb6-9570-f6238bbf2802' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;PAX5&#x27;, &#x27;CD3&#x27;, &#x27;CD11b&#x27;, &#x27;CD11c&#x27;, &#x27;CD15&#x27;, &#x27;CD68&#x27;, &#x27;Podoplanin&#x27;, &#x27;CD31&#x27;,
       &#x27;CD34&#x27;, &#x27;CD90&#x27;, &#x27;CD56&#x27;, &#x27;CD4&#x27;, &#x27;CD8&#x27;, &#x27;Ki-67&#x27;, &#x27;CD45RA&#x27;, &#x27;CD45RO&#x27;],
      dtype=&#x27;object&#x27;, name=&#x27;channels&#x27;))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>features</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-60d0867b-928d-46c3-8720-9eb3c8daa175' class='xr-index-data-in' type='checkbox'/><label for='index-60d0867b-928d-46c3-8720-9eb3c8daa175' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;_labels&#x27;, &#x27;centroid-0&#x27;, &#x27;centroid-1&#x27;], dtype=&#x27;object&#x27;, name=&#x27;features&#x27;))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>la_props</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-fee4605d-aa33-4a73-83eb-6f8f6d73ec3b' class='xr-index-data-in' type='checkbox'/><label for='index-fee4605d-aa33-4a73-83eb-6f8f6d73ec3b' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;_color&#x27;, &#x27;_name&#x27;], dtype=&#x27;object&#x27;, name=&#x27;la_props&#x27;))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>labels</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-254fb360-4ad4-4018-a5f0-0de601c42ab5' class='xr-index-data-in' type='checkbox'/><label for='index-254fb360-4ad4-4018-a5f0-0de601c42ab5' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([1, 2, 3, 4, 5, 6, 7, 8], dtype=&#x27;int64&#x27;, name=&#x27;labels&#x27;))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>x</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-d81f23f7-b440-442a-934e-c33b9952cef0' class='xr-index-data-in' type='checkbox'/><label for='index-d81f23f7-b440-442a-934e-c33b9952cef0' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([   0,    1,    2,    3,    4,    5,    6,    7,    8,    9,
       ...
       2990, 2991, 2992, 2993, 2994, 2995, 2996, 2997, 2998, 2999],
      dtype=&#x27;int64&#x27;, name=&#x27;x&#x27;, length=3000))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>y</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-3af1685d-dfa1-4b59-b4d6-d564e7b75d8c' class='xr-index-data-in' type='checkbox'/><label for='index-3af1685d-dfa1-4b59-b4d6-d564e7b75d8c' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([   0,    1,    2,    3,    4,    5,    6,    7,    8,    9,
       ...
       2990, 2991, 2992, 2993, 2994, 2995, 2996, 2997, 2998, 2999],
      dtype=&#x27;int64&#x27;, name=&#x27;y&#x27;, length=3000))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-6f04c6a5-3e53-4757-84c4-5389b9780223' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-6f04c6a5-3e53-4757-84c4-5389b9780223' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># minor reformatting</span>
<span class="n">seg</span> <span class="o">=</span> <span class="n">ds_processed</span><span class="p">[</span><span class="s2">&quot;_segmentation&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">ds_for_astir</span> <span class="o">=</span> <span class="n">ds_processed</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">drop_layers</span><span class="p">(</span><span class="s2">&quot;_obs&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">drop_layers</span><span class="p">(</span><span class="s2">&quot;_la_properties&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">add_segmentation</span><span class="p">(</span><span class="n">seg</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ds_for_astir</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div><svg style="position: absolute; width: 0; height: 0; overflow: hidden">
<defs>
<symbol id="icon-database" viewBox="0 0 32 32">
<path d="M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z"></path>
<path d="M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
<path d="M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
</symbol>
<symbol id="icon-file-text2" viewBox="0 0 32 32">
<path d="M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z"></path>
<path d="M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
</symbol>
</defs>
</svg>
<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.
 *
 */

:root {
  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));
  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));
  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));
  --xr-border-color: var(--jp-border-color2, #e0e0e0);
  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);
  --xr-background-color: var(--jp-layout-color0, white);
  --xr-background-color-row-even: var(--jp-layout-color1, white);
  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);
}

html[theme=dark],
body[data-theme=dark],
body.vscode-dark {
  --xr-font-color0: rgba(255, 255, 255, 1);
  --xr-font-color2: rgba(255, 255, 255, 0.54);
  --xr-font-color3: rgba(255, 255, 255, 0.38);
  --xr-border-color: #1F1F1F;
  --xr-disabled-color: #515151;
  --xr-background-color: #111111;
  --xr-background-color-row-even: #111111;
  --xr-background-color-row-odd: #313131;
}

.xr-wrap {
  display: block !important;
  min-width: 300px;
  max-width: 700px;
}

.xr-text-repr-fallback {
  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */
  display: none;
}

.xr-header {
  padding-top: 6px;
  padding-bottom: 6px;
  margin-bottom: 4px;
  border-bottom: solid 1px var(--xr-border-color);
}

.xr-header > div,
.xr-header > ul {
  display: inline;
  margin-top: 0;
  margin-bottom: 0;
}

.xr-obj-type,
.xr-array-name {
  margin-left: 2px;
  margin-right: 10px;
}

.xr-obj-type {
  color: var(--xr-font-color2);
}

.xr-sections {
  padding-left: 0 !important;
  display: grid;
  grid-template-columns: 150px auto auto 1fr 20px 20px;
}

.xr-section-item {
  display: contents;
}

.xr-section-item input {
  display: none;
}

.xr-section-item input + label {
  color: var(--xr-disabled-color);
}

.xr-section-item input:enabled + label {
  cursor: pointer;
  color: var(--xr-font-color2);
}

.xr-section-item input:enabled + label:hover {
  color: var(--xr-font-color0);
}

.xr-section-summary {
  grid-column: 1;
  color: var(--xr-font-color2);
  font-weight: 500;
}

.xr-section-summary > span {
  display: inline-block;
  padding-left: 0.5em;
}

.xr-section-summary-in:disabled + label {
  color: var(--xr-font-color2);
}

.xr-section-summary-in + label:before {
  display: inline-block;
  content: '►';
  font-size: 11px;
  width: 15px;
  text-align: center;
}

.xr-section-summary-in:disabled + label:before {
  color: var(--xr-disabled-color);
}

.xr-section-summary-in:checked + label:before {
  content: '▼';
}

.xr-section-summary-in:checked + label > span {
  display: none;
}

.xr-section-summary,
.xr-section-inline-details {
  padding-top: 4px;
  padding-bottom: 4px;
}

.xr-section-inline-details {
  grid-column: 2 / -1;
}

.xr-section-details {
  display: none;
  grid-column: 1 / -1;
  margin-bottom: 5px;
}

.xr-section-summary-in:checked ~ .xr-section-details {
  display: contents;
}

.xr-array-wrap {
  grid-column: 1 / -1;
  display: grid;
  grid-template-columns: 20px auto;
}

.xr-array-wrap > label {
  grid-column: 1;
  vertical-align: top;
}

.xr-preview {
  color: var(--xr-font-color3);
}

.xr-array-preview,
.xr-array-data {
  padding: 0 5px !important;
  grid-column: 2;
}

.xr-array-data,
.xr-array-in:checked ~ .xr-array-preview {
  display: none;
}

.xr-array-in:checked ~ .xr-array-data,
.xr-array-preview {
  display: inline-block;
}

.xr-dim-list {
  display: inline-block !important;
  list-style: none;
  padding: 0 !important;
  margin: 0;
}

.xr-dim-list li {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.xr-dim-list:before {
  content: '(';
}

.xr-dim-list:after {
  content: ')';
}

.xr-dim-list li:not(:last-child):after {
  content: ',';
  padding-right: 5px;
}

.xr-has-index {
  font-weight: bold;
}

.xr-var-list,
.xr-var-item {
  display: contents;
}

.xr-var-item > div,
.xr-var-item label,
.xr-var-item > .xr-var-name span {
  background-color: var(--xr-background-color-row-even);
  margin-bottom: 0;
}

.xr-var-item > .xr-var-name:hover span {
  padding-right: 5px;
}

.xr-var-list > li:nth-child(odd) > div,
.xr-var-list > li:nth-child(odd) > label,
.xr-var-list > li:nth-child(odd) > .xr-var-name span {
  background-color: var(--xr-background-color-row-odd);
}

.xr-var-name {
  grid-column: 1;
}

.xr-var-dims {
  grid-column: 2;
}

.xr-var-dtype {
  grid-column: 3;
  text-align: right;
  color: var(--xr-font-color2);
}

.xr-var-preview {
  grid-column: 4;
}

.xr-index-preview {
  grid-column: 2 / 5;
  color: var(--xr-font-color2);
}

.xr-var-name,
.xr-var-dims,
.xr-var-dtype,
.xr-preview,
.xr-attrs dt {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding-right: 10px;
}

.xr-var-name:hover,
.xr-var-dims:hover,
.xr-var-dtype:hover,
.xr-attrs dt:hover {
  overflow: visible;
  width: auto;
  z-index: 1;
}

.xr-var-attrs,
.xr-var-data,
.xr-index-data {
  display: none;
  background-color: var(--xr-background-color) !important;
  padding-bottom: 5px !important;
}

.xr-var-attrs-in:checked ~ .xr-var-attrs,
.xr-var-data-in:checked ~ .xr-var-data,
.xr-index-data-in:checked ~ .xr-index-data {
  display: block;
}

.xr-var-data > table {
  float: right;
}

.xr-var-name span,
.xr-var-data,
.xr-index-name div,
.xr-index-data,
.xr-attrs {
  padding-left: 25px !important;
}

.xr-attrs,
.xr-var-attrs,
.xr-var-data,
.xr-index-data {
  grid-column: 1 / -1;
}

dl.xr-attrs {
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: 125px auto;
}

.xr-attrs dt,
.xr-attrs dd {
  padding: 0;
  margin: 0;
  float: left;
  padding-right: 10px;
  width: auto;
}

.xr-attrs dt {
  font-weight: normal;
  grid-column: 1;
}

.xr-attrs dt:hover span {
  display: inline-block;
  background: var(--xr-background-color);
  padding-right: 10px;
}

.xr-attrs dd {
  grid-column: 2;
  white-space: pre-wrap;
  word-break: break-all;
}

.xr-icon-database,
.xr-icon-file-text2,
.xr-no-icon {
  display: inline-block;
  vertical-align: middle;
  width: 1em;
  height: 1.5em !important;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
}
</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;
Dimensions:        (channels: 16, x: 3000, y: 3000, cells: 12560, features: 2)
Coordinates:
  * channels       (channels) &lt;U11 &#x27;PAX5&#x27; &#x27;CD3&#x27; &#x27;CD11b&#x27; ... &#x27;CD45RA&#x27; &#x27;CD45RO&#x27;
  * x              (x) int64 0 1 2 3 4 5 6 ... 2994 2995 2996 2997 2998 2999
  * y              (y) int64 0 1 2 3 4 5 6 ... 2994 2995 2996 2997 2998 2999
  * cells          (cells) int64 1 2 3 4 5 6 ... 12556 12557 12558 12559 12560
  * features       (features) &lt;U10 &#x27;centroid-0&#x27; &#x27;centroid-1&#x27;
Data variables:
    _image         (channels, y, x) uint8 0 0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0
    _intensity     (cells, channels) float64 0.004255 2.46 ... 1.384 0.06662
    _segmentation  (y, x) int64 0 0 0 0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0 0 0 0
    _obs           (cells, features) float64 613.3 768.4 ... 2.249e+03 2.237e+03</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-50c15e19-1471-4f56-a7e7-3eae32374199' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-50c15e19-1471-4f56-a7e7-3eae32374199' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span class='xr-has-index'>channels</span>: 16</li><li><span class='xr-has-index'>x</span>: 3000</li><li><span class='xr-has-index'>y</span>: 3000</li><li><span class='xr-has-index'>cells</span>: 12560</li><li><span class='xr-has-index'>features</span>: 2</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-c384cccd-3be7-4473-a1c4-0cf97f226f1a' class='xr-section-summary-in' type='checkbox'  checked><label for='section-c384cccd-3be7-4473-a1c4-0cf97f226f1a' class='xr-section-summary' >Coordinates: <span>(5)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>channels</span></div><div class='xr-var-dims'>(channels)</div><div class='xr-var-dtype'>&lt;U11</div><div class='xr-var-preview xr-preview'>&#x27;PAX5&#x27; &#x27;CD3&#x27; ... &#x27;CD45RA&#x27; &#x27;CD45RO&#x27;</div><input id='attrs-bdcfaef6-f476-48c8-8953-98b2ea6406f9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-bdcfaef6-f476-48c8-8953-98b2ea6406f9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a69c5d16-7ffc-48a7-8319-7d1068272e87' class='xr-var-data-in' type='checkbox'><label for='data-a69c5d16-7ffc-48a7-8319-7d1068272e87' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;PAX5&#x27;, &#x27;CD3&#x27;, &#x27;CD11b&#x27;, &#x27;CD11c&#x27;, &#x27;CD15&#x27;, &#x27;CD68&#x27;, &#x27;Podoplanin&#x27;, &#x27;CD31&#x27;,
       &#x27;CD34&#x27;, &#x27;CD90&#x27;, &#x27;CD56&#x27;, &#x27;CD4&#x27;, &#x27;CD8&#x27;, &#x27;Ki-67&#x27;, &#x27;CD45RA&#x27;, &#x27;CD45RO&#x27;],
      dtype=&#x27;&lt;U11&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>x</span></div><div class='xr-var-dims'>(x)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 4 ... 2996 2997 2998 2999</div><input id='attrs-e1d540f6-04eb-4826-ad80-cc68d5b17925' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e1d540f6-04eb-4826-ad80-cc68d5b17925' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1834103a-5027-49ab-8740-653c79e5bdbb' class='xr-var-data-in' type='checkbox'><label for='data-1834103a-5027-49ab-8740-653c79e5bdbb' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([   0,    1,    2, ..., 2997, 2998, 2999])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>y</span></div><div class='xr-var-dims'>(y)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 4 ... 2996 2997 2998 2999</div><input id='attrs-873dc942-3610-4127-9c5d-08a26961670a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-873dc942-3610-4127-9c5d-08a26961670a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5553f1c8-3679-40bb-b64c-b952c1401b92' class='xr-var-data-in' type='checkbox'><label for='data-5553f1c8-3679-40bb-b64c-b952c1401b92' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([   0,    1,    2, ..., 2997, 2998, 2999])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>cells</span></div><div class='xr-var-dims'>(cells)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>1 2 3 4 ... 12557 12558 12559 12560</div><input id='attrs-c8bbc227-a3ad-4f23-89e0-bd81edc92322' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c8bbc227-a3ad-4f23-89e0-bd81edc92322' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ab265f23-3daf-4be6-aad1-bd99e65ee99d' class='xr-var-data-in' type='checkbox'><label for='data-ab265f23-3daf-4be6-aad1-bd99e65ee99d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([    1,     2,     3, ..., 12558, 12559, 12560])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>features</span></div><div class='xr-var-dims'>(features)</div><div class='xr-var-dtype'>&lt;U10</div><div class='xr-var-preview xr-preview'>&#x27;centroid-0&#x27; &#x27;centroid-1&#x27;</div><input id='attrs-a90d4c3c-412e-4178-8285-927fcb5d3e5e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-a90d4c3c-412e-4178-8285-927fcb5d3e5e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-73b4f5a6-09bb-4284-8d10-81693ec7ec15' class='xr-var-data-in' type='checkbox'><label for='data-73b4f5a6-09bb-4284-8d10-81693ec7ec15' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;centroid-0&#x27;, &#x27;centroid-1&#x27;], dtype=&#x27;&lt;U10&#x27;)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-892e06ce-6b1e-485c-88f0-af59107a4e1f' class='xr-section-summary-in' type='checkbox'  checked><label for='section-892e06ce-6b1e-485c-88f0-af59107a4e1f' class='xr-section-summary' >Data variables: <span>(4)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>_image</span></div><div class='xr-var-dims'>(channels, y, x)</div><div class='xr-var-dtype'>uint8</div><div class='xr-var-preview xr-preview'>0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0</div><input id='attrs-7567e379-dbbb-43f5-981c-da5e4c94b14e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-7567e379-dbbb-43f5-981c-da5e4c94b14e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a8090976-c0a3-489b-a5ce-7ea81a0131cb' class='xr-var-data-in' type='checkbox'><label for='data-a8090976-c0a3-489b-a5ce-7ea81a0131cb' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[[0, 0, 0, ..., 0, 0, 1],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
...
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]]], dtype=uint8)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>_intensity</span></div><div class='xr-var-dims'>(cells, channels)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.004255 2.46 ... 1.384 0.06662</div><input id='attrs-d75934b4-7cc1-48eb-a3d3-86cc310cd344' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d75934b4-7cc1-48eb-a3d3-86cc310cd344' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-00db5bc8-1fd8-4206-9f55-bfec7935fc56' class='xr-var-data-in' type='checkbox'><label for='data-00db5bc8-1fd8-4206-9f55-bfec7935fc56' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[4.25530631e-03, 2.46034155e+00, 2.12765797e-03, ...,
        1.06382959e-03, 4.78540739e-02, 0.00000000e+00],
       [0.00000000e+00, 3.03109280e-01, 1.25937259e+00, ...,
        3.31545474e-02, 0.00000000e+00, 0.00000000e+00],
       [0.00000000e+00, 2.15362817e+00, 9.93224357e-01, ...,
        3.40908431e-03, 0.00000000e+00, 1.11464514e+00],
       ...,
       [1.06773747e-01, 1.35232153e+00, 2.79033557e-02, ...,
        0.00000000e+00, 5.15608185e-01, 6.36717718e-01],
       [1.18304095e+00, 1.13947789e+00, 9.25715906e-02, ...,
        1.71673735e-03, 8.79551538e-01, 2.57510445e-03],
       [8.97647445e-01, 6.25145117e-01, 1.66817809e-01, ...,
        0.00000000e+00, 1.38367696e+00, 6.66173825e-02]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>_segmentation</span></div><div class='xr-var-dims'>(y, x)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0</div><input id='attrs-13354f54-d2a5-47bd-a58d-933b1daea0a0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-13354f54-d2a5-47bd-a58d-933b1daea0a0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7191cb2f-d254-4667-8c81-138fd96a801a' class='xr-var-data-in' type='checkbox'><label for='data-7191cb2f-d254-4667-8c81-138fd96a801a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>_obs</span></div><div class='xr-var-dims'>(cells, features)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>613.3 768.4 ... 2.249e+03 2.237e+03</div><input id='attrs-e7291fb8-8ed2-4a17-904a-d85d754788d7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e7291fb8-8ed2-4a17-904a-d85d754788d7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b85936e6-a9c2-406e-aafa-6dfd6a7d7cb1' class='xr-var-data-in' type='checkbox'><label for='data-b85936e6-a9c2-406e-aafa-6dfd6a7d7cb1' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 613.32978723,  768.42021277],
       [ 769.0984456 ,  707.54404145],
       [ 774.52840909,  644.28409091],
       ...,
       [2265.91085271, 2231.99224806],
       [2281.4806867 , 2218.4248927 ],
       [2249.06944444, 2236.56481481]])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-75aa27c3-2cd8-4388-89f5-8df57270cc40' class='xr-section-summary-in' type='checkbox'  ><label for='section-75aa27c3-2cd8-4388-89f5-8df57270cc40' class='xr-section-summary' >Indexes: <span>(5)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>channels</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-50138c12-6a17-411d-bae6-35c5a9e3b087' class='xr-index-data-in' type='checkbox'/><label for='index-50138c12-6a17-411d-bae6-35c5a9e3b087' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;PAX5&#x27;, &#x27;CD3&#x27;, &#x27;CD11b&#x27;, &#x27;CD11c&#x27;, &#x27;CD15&#x27;, &#x27;CD68&#x27;, &#x27;Podoplanin&#x27;, &#x27;CD31&#x27;,
       &#x27;CD34&#x27;, &#x27;CD90&#x27;, &#x27;CD56&#x27;, &#x27;CD4&#x27;, &#x27;CD8&#x27;, &#x27;Ki-67&#x27;, &#x27;CD45RA&#x27;, &#x27;CD45RO&#x27;],
      dtype=&#x27;object&#x27;, name=&#x27;channels&#x27;))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>x</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-1ae56106-7f75-4c2b-be7b-58c6843234ca' class='xr-index-data-in' type='checkbox'/><label for='index-1ae56106-7f75-4c2b-be7b-58c6843234ca' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([   0,    1,    2,    3,    4,    5,    6,    7,    8,    9,
       ...
       2990, 2991, 2992, 2993, 2994, 2995, 2996, 2997, 2998, 2999],
      dtype=&#x27;int64&#x27;, name=&#x27;x&#x27;, length=3000))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>y</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-af292e5a-e1ed-4d5d-a1a8-066a315e599c' class='xr-index-data-in' type='checkbox'/><label for='index-af292e5a-e1ed-4d5d-a1a8-066a315e599c' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([   0,    1,    2,    3,    4,    5,    6,    7,    8,    9,
       ...
       2990, 2991, 2992, 2993, 2994, 2995, 2996, 2997, 2998, 2999],
      dtype=&#x27;int64&#x27;, name=&#x27;y&#x27;, length=3000))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>cells</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-a27212de-c1bc-49d8-8a34-e993108b53a4' class='xr-index-data-in' type='checkbox'/><label for='index-a27212de-c1bc-49d8-8a34-e993108b53a4' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([    1,     2,     3,     4,     5,     6,     7,     8,     9,    10,
       ...
       12551, 12552, 12553, 12554, 12555, 12556, 12557, 12558, 12559, 12560],
      dtype=&#x27;int64&#x27;, name=&#x27;cells&#x27;, length=12560))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>features</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-021c6ca5-679b-41d1-b806-27c92d8c84fe' class='xr-index-data-in' type='checkbox'/><label for='index-021c6ca5-679b-41d1-b806-27c92d8c84fe' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;centroid-0&#x27;, &#x27;centroid-1&#x27;], dtype=&#x27;object&#x27;, name=&#x27;features&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-2ea4d4b3-3a6f-4772-9421-1df0e52f761d' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-2ea4d4b3-3a6f-4772-9421-1df0e52f761d' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># using astir to predict cell types</span>
<span class="c1"># note the slightly different structure of the marker dictionary</span>
<span class="n">ct_marker_dict</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;cell_type&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;PAX5&quot;</span><span class="p">],</span>
        <span class="s2">&quot;T&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CD3&quot;</span><span class="p">],</span>
        <span class="s2">&quot;Myeloid&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CD11b&quot;</span><span class="p">],</span>
        <span class="s2">&quot;Dendritic&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CD11c&quot;</span><span class="p">],</span>
        <span class="s2">&quot;Granulo&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CD15&quot;</span><span class="p">],</span>
        <span class="s2">&quot;Macro&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CD68&quot;</span><span class="p">],</span>
        <span class="s2">&quot;Stroma PDPN&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Podoplanin&quot;</span><span class="p">],</span>
        <span class="s2">&quot;Stroma CD31&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CD31&quot;</span><span class="p">],</span>
        <span class="s2">&quot;Stroma CD34&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CD34&quot;</span><span class="p">],</span>
        <span class="s2">&quot;Stroma CD90&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CD90&quot;</span><span class="p">],</span>
        <span class="s2">&quot;NK&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CD56&quot;</span><span class="p">],</span>
    <span class="p">}</span>
<span class="p">}</span>
<span class="n">ds_with_ct_predictions_astir</span> <span class="o">=</span> <span class="n">ds_for_astir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">astir</span><span class="p">(</span><span class="n">ct_marker_dict</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
training restart 1/5: 100%|██████████| 5/5 [ 3.78s/epochs, current loss: -17157.1]
training restart 2/5: 100%|██████████| 5/5 [ 3.28s/epochs, current loss: -18852.9]
training restart 3/5: 100%|██████████| 5/5 [ 3.30s/epochs, current loss: -19201.0]
training restart 4/5: 100%|██████████| 5/5 [ 3.28s/epochs, current loss: -17167.5]
training restart 5/5: 100%|██████████| 5/5 [ 3.42s/epochs, current loss: -17890.4]
training restart (final):   4%|▍         | 21/500 [ 3.86s/epochs, current loss: -47704.8]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># adding custom colors and adding observations back in</span>
<span class="n">cell_types</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;B&quot;</span><span class="p">,</span>
    <span class="s2">&quot;T&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Myeloid&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Dendritic&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Granulo&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Macro&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Stroma PDPN&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Stroma CD31&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Stroma CD34&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Stroma CD90&quot;</span><span class="p">,</span>
    <span class="s2">&quot;NK&quot;</span><span class="p">,</span>
    <span class="s2">&quot;Other&quot;</span><span class="p">,</span>
<span class="p">]</span>
<span class="n">ds_with_ct_predictions_astir</span> <span class="o">=</span> <span class="n">ds_with_ct_predictions_astir</span><span class="o">.</span><span class="n">la</span><span class="o">.</span><span class="n">set_label_colors</span><span class="p">(</span>
    <span class="n">cell_types</span><span class="p">,</span> <span class="n">colors</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;darkgray&quot;</span><span class="p">]</span>
<span class="p">)</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">add_observations</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ds_with_ct_predictions_astir</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div><svg style="position: absolute; width: 0; height: 0; overflow: hidden">
<defs>
<symbol id="icon-database" viewBox="0 0 32 32">
<path d="M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z"></path>
<path d="M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
<path d="M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
</symbol>
<symbol id="icon-file-text2" viewBox="0 0 32 32">
<path d="M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z"></path>
<path d="M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
</symbol>
</defs>
</svg>
<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.
 *
 */

:root {
  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));
  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));
  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));
  --xr-border-color: var(--jp-border-color2, #e0e0e0);
  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);
  --xr-background-color: var(--jp-layout-color0, white);
  --xr-background-color-row-even: var(--jp-layout-color1, white);
  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);
}

html[theme=dark],
body[data-theme=dark],
body.vscode-dark {
  --xr-font-color0: rgba(255, 255, 255, 1);
  --xr-font-color2: rgba(255, 255, 255, 0.54);
  --xr-font-color3: rgba(255, 255, 255, 0.38);
  --xr-border-color: #1F1F1F;
  --xr-disabled-color: #515151;
  --xr-background-color: #111111;
  --xr-background-color-row-even: #111111;
  --xr-background-color-row-odd: #313131;
}

.xr-wrap {
  display: block !important;
  min-width: 300px;
  max-width: 700px;
}

.xr-text-repr-fallback {
  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */
  display: none;
}

.xr-header {
  padding-top: 6px;
  padding-bottom: 6px;
  margin-bottom: 4px;
  border-bottom: solid 1px var(--xr-border-color);
}

.xr-header > div,
.xr-header > ul {
  display: inline;
  margin-top: 0;
  margin-bottom: 0;
}

.xr-obj-type,
.xr-array-name {
  margin-left: 2px;
  margin-right: 10px;
}

.xr-obj-type {
  color: var(--xr-font-color2);
}

.xr-sections {
  padding-left: 0 !important;
  display: grid;
  grid-template-columns: 150px auto auto 1fr 20px 20px;
}

.xr-section-item {
  display: contents;
}

.xr-section-item input {
  display: none;
}

.xr-section-item input + label {
  color: var(--xr-disabled-color);
}

.xr-section-item input:enabled + label {
  cursor: pointer;
  color: var(--xr-font-color2);
}

.xr-section-item input:enabled + label:hover {
  color: var(--xr-font-color0);
}

.xr-section-summary {
  grid-column: 1;
  color: var(--xr-font-color2);
  font-weight: 500;
}

.xr-section-summary > span {
  display: inline-block;
  padding-left: 0.5em;
}

.xr-section-summary-in:disabled + label {
  color: var(--xr-font-color2);
}

.xr-section-summary-in + label:before {
  display: inline-block;
  content: '►';
  font-size: 11px;
  width: 15px;
  text-align: center;
}

.xr-section-summary-in:disabled + label:before {
  color: var(--xr-disabled-color);
}

.xr-section-summary-in:checked + label:before {
  content: '▼';
}

.xr-section-summary-in:checked + label > span {
  display: none;
}

.xr-section-summary,
.xr-section-inline-details {
  padding-top: 4px;
  padding-bottom: 4px;
}

.xr-section-inline-details {
  grid-column: 2 / -1;
}

.xr-section-details {
  display: none;
  grid-column: 1 / -1;
  margin-bottom: 5px;
}

.xr-section-summary-in:checked ~ .xr-section-details {
  display: contents;
}

.xr-array-wrap {
  grid-column: 1 / -1;
  display: grid;
  grid-template-columns: 20px auto;
}

.xr-array-wrap > label {
  grid-column: 1;
  vertical-align: top;
}

.xr-preview {
  color: var(--xr-font-color3);
}

.xr-array-preview,
.xr-array-data {
  padding: 0 5px !important;
  grid-column: 2;
}

.xr-array-data,
.xr-array-in:checked ~ .xr-array-preview {
  display: none;
}

.xr-array-in:checked ~ .xr-array-data,
.xr-array-preview {
  display: inline-block;
}

.xr-dim-list {
  display: inline-block !important;
  list-style: none;
  padding: 0 !important;
  margin: 0;
}

.xr-dim-list li {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.xr-dim-list:before {
  content: '(';
}

.xr-dim-list:after {
  content: ')';
}

.xr-dim-list li:not(:last-child):after {
  content: ',';
  padding-right: 5px;
}

.xr-has-index {
  font-weight: bold;
}

.xr-var-list,
.xr-var-item {
  display: contents;
}

.xr-var-item > div,
.xr-var-item label,
.xr-var-item > .xr-var-name span {
  background-color: var(--xr-background-color-row-even);
  margin-bottom: 0;
}

.xr-var-item > .xr-var-name:hover span {
  padding-right: 5px;
}

.xr-var-list > li:nth-child(odd) > div,
.xr-var-list > li:nth-child(odd) > label,
.xr-var-list > li:nth-child(odd) > .xr-var-name span {
  background-color: var(--xr-background-color-row-odd);
}

.xr-var-name {
  grid-column: 1;
}

.xr-var-dims {
  grid-column: 2;
}

.xr-var-dtype {
  grid-column: 3;
  text-align: right;
  color: var(--xr-font-color2);
}

.xr-var-preview {
  grid-column: 4;
}

.xr-index-preview {
  grid-column: 2 / 5;
  color: var(--xr-font-color2);
}

.xr-var-name,
.xr-var-dims,
.xr-var-dtype,
.xr-preview,
.xr-attrs dt {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding-right: 10px;
}

.xr-var-name:hover,
.xr-var-dims:hover,
.xr-var-dtype:hover,
.xr-attrs dt:hover {
  overflow: visible;
  width: auto;
  z-index: 1;
}

.xr-var-attrs,
.xr-var-data,
.xr-index-data {
  display: none;
  background-color: var(--xr-background-color) !important;
  padding-bottom: 5px !important;
}

.xr-var-attrs-in:checked ~ .xr-var-attrs,
.xr-var-data-in:checked ~ .xr-var-data,
.xr-index-data-in:checked ~ .xr-index-data {
  display: block;
}

.xr-var-data > table {
  float: right;
}

.xr-var-name span,
.xr-var-data,
.xr-index-name div,
.xr-index-data,
.xr-attrs {
  padding-left: 25px !important;
}

.xr-attrs,
.xr-var-attrs,
.xr-var-data,
.xr-index-data {
  grid-column: 1 / -1;
}

dl.xr-attrs {
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: 125px auto;
}

.xr-attrs dt,
.xr-attrs dd {
  padding: 0;
  margin: 0;
  float: left;
  padding-right: 10px;
  width: auto;
}

.xr-attrs dt {
  font-weight: normal;
  grid-column: 1;
}

.xr-attrs dt:hover span {
  display: inline-block;
  background: var(--xr-background-color);
  padding-right: 10px;
}

.xr-attrs dd {
  grid-column: 2;
  white-space: pre-wrap;
  word-break: break-all;
}

.xr-icon-database,
.xr-icon-file-text2,
.xr-no-icon {
  display: inline-block;
  vertical-align: middle;
  width: 1em;
  height: 1.5em !important;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
}
</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;
Dimensions:         (labels: 12, la_props: 2, channels: 16, x: 3000, y: 3000,
                     cells: 12560, features: 3)
Coordinates:
  * labels          (labels) int64 1 2 3 4 5 6 7 8 9 10 11 12
  * la_props        (la_props) &lt;U6 &#x27;_color&#x27; &#x27;_name&#x27;
  * channels        (channels) &lt;U11 &#x27;PAX5&#x27; &#x27;CD3&#x27; &#x27;CD11b&#x27; ... &#x27;CD45RA&#x27; &#x27;CD45RO&#x27;
  * x               (x) int64 0 1 2 3 4 5 6 ... 2994 2995 2996 2997 2998 2999
  * y               (y) int64 0 1 2 3 4 5 6 ... 2994 2995 2996 2997 2998 2999
  * cells           (cells) int64 1 2 3 4 5 6 ... 12556 12557 12558 12559 12560
  * features        (features) &lt;U10 &#x27;_labels&#x27; &#x27;centroid-0&#x27; &#x27;centroid-1&#x27;
Data variables:
    _image          (channels, y, x) uint8 0 0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0
    _intensity      (cells, channels) float64 0.004255 2.46 ... 1.384 0.06662
    _segmentation   (y, x) int64 0 0 0 0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0 0 0 0
    _obs            (cells, features) float64 12.0 613.3 ... 2.249e+03 2.237e+03
    _la_properties  (labels, la_props) object &#x27;#e6194B&#x27; &#x27;B&#x27; ... &#x27;#3cb44b&#x27; &#x27;T&#x27;</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-dd4dabde-db6c-476f-bf56-2cdd39ba0caf' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-dd4dabde-db6c-476f-bf56-2cdd39ba0caf' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span class='xr-has-index'>labels</span>: 12</li><li><span class='xr-has-index'>la_props</span>: 2</li><li><span class='xr-has-index'>channels</span>: 16</li><li><span class='xr-has-index'>x</span>: 3000</li><li><span class='xr-has-index'>y</span>: 3000</li><li><span class='xr-has-index'>cells</span>: 12560</li><li><span class='xr-has-index'>features</span>: 3</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-16a29ba7-dd80-4865-80e3-8706580b2ed2' class='xr-section-summary-in' type='checkbox'  checked><label for='section-16a29ba7-dd80-4865-80e3-8706580b2ed2' class='xr-section-summary' >Coordinates: <span>(7)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>labels</span></div><div class='xr-var-dims'>(labels)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>1 2 3 4 5 6 7 8 9 10 11 12</div><input id='attrs-e4a25635-99ef-4877-97a1-22d59e6dc9b8' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e4a25635-99ef-4877-97a1-22d59e6dc9b8' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-16a0dcb5-d472-4d97-af76-50c5a8a62f0a' class='xr-var-data-in' type='checkbox'><label for='data-16a0dcb5-d472-4d97-af76-50c5a8a62f0a' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>la_props</span></div><div class='xr-var-dims'>(la_props)</div><div class='xr-var-dtype'>&lt;U6</div><div class='xr-var-preview xr-preview'>&#x27;_color&#x27; &#x27;_name&#x27;</div><input id='attrs-4cbaf8b3-db3c-44e8-9564-448567dbc94e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-4cbaf8b3-db3c-44e8-9564-448567dbc94e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9d0887ab-dafd-47db-bd79-60b5ad7b2f91' class='xr-var-data-in' type='checkbox'><label for='data-9d0887ab-dafd-47db-bd79-60b5ad7b2f91' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;_color&#x27;, &#x27;_name&#x27;], dtype=&#x27;&lt;U6&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>channels</span></div><div class='xr-var-dims'>(channels)</div><div class='xr-var-dtype'>&lt;U11</div><div class='xr-var-preview xr-preview'>&#x27;PAX5&#x27; &#x27;CD3&#x27; ... &#x27;CD45RA&#x27; &#x27;CD45RO&#x27;</div><input id='attrs-41def3c9-93f5-43e2-a04c-13e36c532a04' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-41def3c9-93f5-43e2-a04c-13e36c532a04' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b30b8107-52fa-41aa-ab56-e1e2917dbbdc' class='xr-var-data-in' type='checkbox'><label for='data-b30b8107-52fa-41aa-ab56-e1e2917dbbdc' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;PAX5&#x27;, &#x27;CD3&#x27;, &#x27;CD11b&#x27;, &#x27;CD11c&#x27;, &#x27;CD15&#x27;, &#x27;CD68&#x27;, &#x27;Podoplanin&#x27;, &#x27;CD31&#x27;,
       &#x27;CD34&#x27;, &#x27;CD90&#x27;, &#x27;CD56&#x27;, &#x27;CD4&#x27;, &#x27;CD8&#x27;, &#x27;Ki-67&#x27;, &#x27;CD45RA&#x27;, &#x27;CD45RO&#x27;],
      dtype=&#x27;&lt;U11&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>x</span></div><div class='xr-var-dims'>(x)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 4 ... 2996 2997 2998 2999</div><input id='attrs-888db18a-e9ac-479b-9182-b6021e450d0e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-888db18a-e9ac-479b-9182-b6021e450d0e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-03269ec1-198b-4cd9-813c-524bad654bd3' class='xr-var-data-in' type='checkbox'><label for='data-03269ec1-198b-4cd9-813c-524bad654bd3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([   0,    1,    2, ..., 2997, 2998, 2999])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>y</span></div><div class='xr-var-dims'>(y)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 4 ... 2996 2997 2998 2999</div><input id='attrs-ae82636d-89b8-4b83-8a7c-4bd83d582448' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ae82636d-89b8-4b83-8a7c-4bd83d582448' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a6d82243-9138-4557-af5a-9acc354d6281' class='xr-var-data-in' type='checkbox'><label for='data-a6d82243-9138-4557-af5a-9acc354d6281' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([   0,    1,    2, ..., 2997, 2998, 2999])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>cells</span></div><div class='xr-var-dims'>(cells)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>1 2 3 4 ... 12557 12558 12559 12560</div><input id='attrs-892c62b8-d03d-44ba-af99-e1b7e4af49f9' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-892c62b8-d03d-44ba-af99-e1b7e4af49f9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-57ab40ae-9b44-4b48-890d-65add3497086' class='xr-var-data-in' type='checkbox'><label for='data-57ab40ae-9b44-4b48-890d-65add3497086' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([    1,     2,     3, ..., 12558, 12559, 12560])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>features</span></div><div class='xr-var-dims'>(features)</div><div class='xr-var-dtype'>&lt;U10</div><div class='xr-var-preview xr-preview'>&#x27;_labels&#x27; &#x27;centroid-0&#x27; &#x27;centroid-1&#x27;</div><input id='attrs-eb743e37-da9a-4119-872a-d68247f8c8ea' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-eb743e37-da9a-4119-872a-d68247f8c8ea' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-75fcac3a-1b19-4be2-a436-7f68271da83f' class='xr-var-data-in' type='checkbox'><label for='data-75fcac3a-1b19-4be2-a436-7f68271da83f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;_labels&#x27;, &#x27;centroid-0&#x27;, &#x27;centroid-1&#x27;], dtype=&#x27;&lt;U10&#x27;)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-f5435a8c-ee6b-4357-88b0-2b765a04972c' class='xr-section-summary-in' type='checkbox'  checked><label for='section-f5435a8c-ee6b-4357-88b0-2b765a04972c' class='xr-section-summary' >Data variables: <span>(5)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>_image</span></div><div class='xr-var-dims'>(channels, y, x)</div><div class='xr-var-dtype'>uint8</div><div class='xr-var-preview xr-preview'>0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0</div><input id='attrs-e948e967-f900-457f-af88-16b3d20927ac' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e948e967-f900-457f-af88-16b3d20927ac' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5ef76eb1-2cde-447e-933a-ce123edd73e7' class='xr-var-data-in' type='checkbox'><label for='data-5ef76eb1-2cde-447e-933a-ce123edd73e7' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[[0, 0, 0, ..., 0, 0, 1],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
...
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]]], dtype=uint8)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>_intensity</span></div><div class='xr-var-dims'>(cells, channels)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.004255 2.46 ... 1.384 0.06662</div><input id='attrs-cf46b1cd-622c-42b0-aefb-be6b55a17a5d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-cf46b1cd-622c-42b0-aefb-be6b55a17a5d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-34770cc9-7820-4780-ac1e-5e9b8c30c4c6' class='xr-var-data-in' type='checkbox'><label for='data-34770cc9-7820-4780-ac1e-5e9b8c30c4c6' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[4.25530631e-03, 2.46034155e+00, 2.12765797e-03, ...,
        1.06382959e-03, 4.78540739e-02, 0.00000000e+00],
       [0.00000000e+00, 3.03109280e-01, 1.25937259e+00, ...,
        3.31545474e-02, 0.00000000e+00, 0.00000000e+00],
       [0.00000000e+00, 2.15362817e+00, 9.93224357e-01, ...,
        3.40908431e-03, 0.00000000e+00, 1.11464514e+00],
       ...,
       [1.06773747e-01, 1.35232153e+00, 2.79033557e-02, ...,
        0.00000000e+00, 5.15608185e-01, 6.36717718e-01],
       [1.18304095e+00, 1.13947789e+00, 9.25715906e-02, ...,
        1.71673735e-03, 8.79551538e-01, 2.57510445e-03],
       [8.97647445e-01, 6.25145117e-01, 1.66817809e-01, ...,
        0.00000000e+00, 1.38367696e+00, 6.66173825e-02]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>_segmentation</span></div><div class='xr-var-dims'>(y, x)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0</div><input id='attrs-033bea3d-79dd-4428-a204-c3afbe0f0bc4' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-033bea3d-79dd-4428-a204-c3afbe0f0bc4' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-4ca6994f-df34-4cca-bd86-0f834a68096d' class='xr-var-data-in' type='checkbox'><label for='data-4ca6994f-df34-4cca-bd86-0f834a68096d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>_obs</span></div><div class='xr-var-dims'>(cells, features)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>12.0 613.3 ... 2.249e+03 2.237e+03</div><input id='attrs-68611a01-029e-4a94-8abf-5e96620f5a32' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-68611a01-029e-4a94-8abf-5e96620f5a32' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-86d3bb89-02e3-48e5-9cac-675a002d85d3' class='xr-var-data-in' type='checkbox'><label for='data-86d3bb89-02e3-48e5-9cac-675a002d85d3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[1.20000000e+01, 6.13329787e+02, 7.68420213e+02],
       [4.00000000e+00, 7.69098446e+02, 7.07544041e+02],
       [9.00000000e+00, 7.74528409e+02, 6.44284091e+02],
       ...,
       [7.00000000e+00, 2.26591085e+03, 2.23199225e+03],
       [1.00000000e+00, 2.28148069e+03, 2.21842489e+03],
       [1.00000000e+00, 2.24906944e+03, 2.23656481e+03]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>_la_properties</span></div><div class='xr-var-dims'>(labels, la_props)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;#e6194B&#x27; &#x27;B&#x27; ... &#x27;#3cb44b&#x27; &#x27;T&#x27;</div><input id='attrs-42e86802-2dc7-4913-a7eb-92378290decc' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-42e86802-2dc7-4913-a7eb-92378290decc' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-4aeb060e-c9de-47b3-8c3c-35b2ed286c93' class='xr-var-data-in' type='checkbox'><label for='data-4aeb060e-c9de-47b3-8c3c-35b2ed286c93' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[&#x27;#e6194B&#x27;, &#x27;B&#x27;],
       [&#x27;#4363d8&#x27;, &#x27;Dendritic&#x27;],
       [&#x27;#ffd8b1&#x27;, &#x27;Granulo&#x27;],
       [&#x27;#f58231&#x27;, &#x27;Macro&#x27;],
       [&#x27;#ffe119&#x27;, &#x27;Myeloid&#x27;],
       [&#x27;#9A6324&#x27;, &#x27;NK&#x27;],
       [&#x27;darkgray&#x27;, &#x27;Other&#x27;],
       [&#x27;#fffac8&#x27;, &#x27;Stroma CD31&#x27;],
       [&#x27;#469990&#x27;, &#x27;Stroma CD34&#x27;],
       [&#x27;#fabed4&#x27;, &#x27;Stroma CD90&#x27;],
       [&#x27;#911eb4&#x27;, &#x27;Stroma PDPN&#x27;],
       [&#x27;#3cb44b&#x27;, &#x27;T&#x27;]], dtype=object)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-1b65c377-abfe-4506-bd38-6c8e7aa8b9f5' class='xr-section-summary-in' type='checkbox'  ><label for='section-1b65c377-abfe-4506-bd38-6c8e7aa8b9f5' class='xr-section-summary' >Indexes: <span>(7)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>labels</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-e103e848-b08c-4e69-bfa8-e64e036cc012' class='xr-index-data-in' type='checkbox'/><label for='index-e103e848-b08c-4e69-bfa8-e64e036cc012' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12], dtype=&#x27;int64&#x27;, name=&#x27;labels&#x27;))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>la_props</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-7a626eda-81d6-471c-b5d4-030e41a3f4ee' class='xr-index-data-in' type='checkbox'/><label for='index-7a626eda-81d6-471c-b5d4-030e41a3f4ee' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;_color&#x27;, &#x27;_name&#x27;], dtype=&#x27;object&#x27;, name=&#x27;la_props&#x27;))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>channels</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-c1db15b9-6dc5-459b-ae7d-897a729d3305' class='xr-index-data-in' type='checkbox'/><label for='index-c1db15b9-6dc5-459b-ae7d-897a729d3305' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;PAX5&#x27;, &#x27;CD3&#x27;, &#x27;CD11b&#x27;, &#x27;CD11c&#x27;, &#x27;CD15&#x27;, &#x27;CD68&#x27;, &#x27;Podoplanin&#x27;, &#x27;CD31&#x27;,
       &#x27;CD34&#x27;, &#x27;CD90&#x27;, &#x27;CD56&#x27;, &#x27;CD4&#x27;, &#x27;CD8&#x27;, &#x27;Ki-67&#x27;, &#x27;CD45RA&#x27;, &#x27;CD45RO&#x27;],
      dtype=&#x27;object&#x27;, name=&#x27;channels&#x27;))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>x</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-a1f12bf0-5e3e-4b66-88e6-d382e7f19e34' class='xr-index-data-in' type='checkbox'/><label for='index-a1f12bf0-5e3e-4b66-88e6-d382e7f19e34' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([   0,    1,    2,    3,    4,    5,    6,    7,    8,    9,
       ...
       2990, 2991, 2992, 2993, 2994, 2995, 2996, 2997, 2998, 2999],
      dtype=&#x27;int64&#x27;, name=&#x27;x&#x27;, length=3000))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>y</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-d20664d6-cdf2-4926-81ba-7799f3cd5932' class='xr-index-data-in' type='checkbox'/><label for='index-d20664d6-cdf2-4926-81ba-7799f3cd5932' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([   0,    1,    2,    3,    4,    5,    6,    7,    8,    9,
       ...
       2990, 2991, 2992, 2993, 2994, 2995, 2996, 2997, 2998, 2999],
      dtype=&#x27;int64&#x27;, name=&#x27;y&#x27;, length=3000))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>cells</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-f02efd78-e25f-4c3b-a6f5-763b91cdcecf' class='xr-index-data-in' type='checkbox'/><label for='index-f02efd78-e25f-4c3b-a6f5-763b91cdcecf' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([    1,     2,     3,     4,     5,     6,     7,     8,     9,    10,
       ...
       12551, 12552, 12553, 12554, 12555, 12556, 12557, 12558, 12559, 12560],
      dtype=&#x27;int64&#x27;, name=&#x27;cells&#x27;, length=12560))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>features</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-f378e801-c59f-48e8-b980-c4ac5d8930e1' class='xr-index-data-in' type='checkbox'/><label for='index-f378e801-c59f-48e8-b980-c4ac5d8930e1' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;_labels&#x27;, &#x27;centroid-0&#x27;, &#x27;centroid-1&#x27;], dtype=&#x27;object&#x27;, name=&#x27;features&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-a4ef6b2d-2b03-4498-bab7-663d7d79079a' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-a4ef6b2d-2b03-4498-bab7-663d7d79079a' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plotting the astir and the argmax predictions</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">21</span><span class="p">,</span> <span class="mi">7</span><span class="p">))</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">ds_processed</span><span class="o">.</span><span class="n">pp</span><span class="p">[</span><span class="n">channels_celltypes</span><span class="p">]</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">colorize</span><span class="p">(</span><span class="n">colors</span><span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">autocrop</span><span class="p">(</span><span class="n">padding</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">ds_with_ct_predictions_astir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">autocrop</span><span class="p">(</span><span class="n">padding</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">render_image</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">render_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">ds_with_ct_predictions</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">autocrop</span><span class="p">(</span><span class="n">padding</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">render_image</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">render_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>

<span class="c1"># removing the x and y ticks</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>

<span class="c1"># setting titles</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Markers&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Astir Predictions&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Argmax Predictions&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_CellTypePrediction_21_0.png" src="../_images/notebooks_CellTypePrediction_21_0.png" />
</div>
</div>
</section>
<section id="Marker-Binarization">
<h2>Marker Binarization<a class="headerlink" href="#Marker-Binarization" title="Link to this heading">¶</a></h2>
<p>For some (functional) markers, we want to binarize them (i. e. is a cell positive or negative for that marker). This is performed in several steps: 1. First, the image needs to be thresholded. We already did this at the beginning of the notebook using the <code class="docutils literal notranslate"><span class="pre">pp.threshold()</span></code> method. 2. Afterwards, we need to once again quantify the expression for each cell. While we have already seen how to do this using the mean expression, we can also use things such as the percentage of positive pixels (after
thresholding). 3. Next, we actually perform the binarization using <code class="docutils literal notranslate"><span class="pre">la.threshold_labels()</span></code>. This way, we could for example consider a cell positive for CD4 only if more than 90% of its pixels have a non-zero intensity value. 4. Finally, we create a hierarchy of cell subtypes, which is based on the labels we computed previously and subsets them further by using the binarized functional markers.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># adding a new quantification: the percentage of positive pixels (this requires appropriate thresholding, as we have performed ealier)</span>
<span class="n">ds_with_ct_predictions</span> <span class="o">=</span> <span class="n">ds_with_ct_predictions</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">add_quantification</span><span class="p">(</span>
    <span class="n">func</span><span class="o">=</span><span class="n">sp</span><span class="o">.</span><span class="n">percentage_positive</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;_percentage_positive&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># these thresholds are the percentages of positive cells required</span>
<span class="c1"># so a cell with more than 50% of non-zero CD4 pixels would be considered CD4 positive</span>
<span class="n">threshold_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;CD4&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s2">&quot;CD8&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s2">&quot;Ki-67&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s2">&quot;CD45RA&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s2">&quot;CD45RO&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">}</span>

<span class="n">ds_with_binarization</span> <span class="o">=</span> <span class="n">ds_with_ct_predictions</span><span class="o">.</span><span class="n">la</span><span class="o">.</span><span class="n">threshold_labels</span><span class="p">(</span><span class="n">threshold_dict</span><span class="p">,</span> <span class="n">layer_key</span><span class="o">=</span><span class="s2">&quot;_percentage_positive&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># under the hood, we can see that our new object now contains binarization columns in the _obs field</span>
<span class="n">ds_with_binarization</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">get_layer_as_df</span><span class="p">()</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CD45RA_binarized</th>
      <th>CD45RO_binarized</th>
      <th>CD4_binarized</th>
      <th>CD8_binarized</th>
      <th>Ki-67_binarized</th>
      <th>_labels</th>
      <th>centroid-0</th>
      <th>centroid-1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>T</td>
      <td>613.329787</td>
      <td>768.420213</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>Macro</td>
      <td>769.098446</td>
      <td>707.544041</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>T</td>
      <td>774.528409</td>
      <td>644.284091</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>T</td>
      <td>775.902878</td>
      <td>592.744604</td>
    </tr>
    <tr>
      <th>5</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>Myeloid</td>
      <td>668.844749</td>
      <td>729.310502</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
<section id="Cell-Subtype-Prediction">
<h2>Cell Subtype Prediction<a class="headerlink" href="#Cell-Subtype-Prediction" title="Link to this heading">¶</a></h2>
<p>Finally, it is time to combine the broad cell type predictions with the binarization results in order to predict more fine-grained cell types. Here, we use a dictionary to define a cell type hierarchy, and then use the method <code class="docutils literal notranslate"><span class="pre">la.predict_cell_subtypes()</span></code> to subset our cell types accordingly. Instead of a dictionary, you can also provide the path to a yaml file. This file could look something like this:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>B:
  subtypes:
    - name: B_prol
      markers: [&quot;Ki67+&quot;]
T:
  subtypes:
    - name: T_h
      markers: [&quot;CD4+&quot;]
    - name: T_tox
      markers: [&quot;CD8+&quot;]
      subtypes:
        - name: T_tox_naive
          markers: [&quot;CD45RA+&quot;]
        - name: T_tox_mem
          markers: [&quot;CD45RO+&quot;]
</pre></div>
</div>
<p>Note that after each marker, we add a <code class="docutils literal notranslate"><span class="pre">+</span></code> after its name if we want cells of this cell type to be positive for a marker, or a <code class="docutils literal notranslate"><span class="pre">-</span></code> if we require it to be negative for the marker instead.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># calling of cell subtypes with a provided hierarchy</span>
<span class="n">subtype_dict</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;subtypes&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;B_prol&quot;</span><span class="p">,</span> <span class="s2">&quot;markers&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Ki-67+&quot;</span><span class="p">]}]},</span>
    <span class="s2">&quot;T&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;subtypes&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="p">{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;T_h&quot;</span><span class="p">,</span> <span class="s2">&quot;markers&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CD4+&quot;</span><span class="p">]},</span>
            <span class="p">{</span>
                <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;T_tox&quot;</span><span class="p">,</span>
                <span class="s2">&quot;markers&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CD8+&quot;</span><span class="p">],</span>
                <span class="s2">&quot;subtypes&quot;</span><span class="p">:</span> <span class="p">[</span>
                    <span class="p">{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;T_tox_naive&quot;</span><span class="p">,</span> <span class="s2">&quot;markers&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CD45RA+&quot;</span><span class="p">]},</span>
                    <span class="p">{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;T_tox_mem&quot;</span><span class="p">,</span> <span class="s2">&quot;markers&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CD45RO+&quot;</span><span class="p">]},</span>
                <span class="p">],</span>
            <span class="p">},</span>
        <span class="p">]</span>
    <span class="p">},</span>
<span class="p">}</span>

<span class="n">ds_with_subtypes</span> <span class="o">=</span> <span class="n">ds_with_binarization</span><span class="o">.</span><span class="n">la</span><span class="o">.</span><span class="n">predict_cell_subtypes</span><span class="p">(</span><span class="n">subtype_dict</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># if you wanted to read the hierarchy from a file, you could do it like this</span>
<span class="n">ds_with_subtypes</span> <span class="o">=</span> <span class="n">ds_with_binarization</span><span class="o">.</span><span class="n">la</span><span class="o">.</span><span class="n">predict_cell_subtypes</span><span class="p">(</span><span class="s2">&quot;../../data/subtype_hierarchy.yaml&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>You can see that this added cell type labels on multiple levels to the <code class="docutils literal notranslate"><span class="pre">_la_layers</span></code> table, so you can always go back to a lower level. In addition, the most granular cell type labels have been added as the new default labels in <code class="docutils literal notranslate"><span class="pre">_obs</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ds_with_subtypes</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">get_layer_as_df</span><span class="p">(</span><span class="s2">&quot;_la_layers&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">15</span><span class="p">:</span><span class="mi">25</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>labels_0</th>
      <th>labels_1</th>
      <th>labels_2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>16</th>
      <td>T</td>
      <td>T_h</td>
      <td>T_h</td>
    </tr>
    <tr>
      <th>17</th>
      <td>Macro</td>
      <td>Macro</td>
      <td>Macro</td>
    </tr>
    <tr>
      <th>18</th>
      <td>T</td>
      <td>T</td>
      <td>T</td>
    </tr>
    <tr>
      <th>19</th>
      <td>T</td>
      <td>T_tox</td>
      <td>T_tox</td>
    </tr>
    <tr>
      <th>20</th>
      <td>T</td>
      <td>T</td>
      <td>T</td>
    </tr>
    <tr>
      <th>21</th>
      <td>T</td>
      <td>T</td>
      <td>T</td>
    </tr>
    <tr>
      <th>22</th>
      <td>T</td>
      <td>T</td>
      <td>T</td>
    </tr>
    <tr>
      <th>23</th>
      <td>T</td>
      <td>T</td>
      <td>T</td>
    </tr>
    <tr>
      <th>24</th>
      <td>T</td>
      <td>T_tox</td>
      <td>T_tox</td>
    </tr>
    <tr>
      <th>25</th>
      <td>T</td>
      <td>T_h</td>
      <td>T_h</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plotting the cell type predictions</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">ds_with_subtypes</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">autocrop</span><span class="p">()</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">render_image</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">render_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_CellTypePrediction_31_0.png" src="../_images/notebooks_CellTypePrediction_31_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plotting the cell type predictions</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">_</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ds_with_subtypes</span><span class="o">.</span><span class="n">la</span><span class="p">[[</span><span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;T_h&quot;</span><span class="p">,</span> <span class="s2">&quot;T_tox&quot;</span><span class="p">,</span> <span class="s2">&quot;T_tox_mem&quot;</span><span class="p">,</span> <span class="s2">&quot;T_tox_naive&quot;</span><span class="p">]]</span>
    <span class="o">.</span><span class="n">la</span><span class="o">.</span><span class="n">set_label_colors</span><span class="p">(</span>
        <span class="p">[</span><span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;T_h&quot;</span><span class="p">,</span> <span class="s2">&quot;T_tox&quot;</span><span class="p">,</span> <span class="s2">&quot;T_tox_mem&quot;</span><span class="p">,</span> <span class="s2">&quot;T_tox_naive&quot;</span><span class="p">],</span>
        <span class="n">colors</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gray&quot;</span><span class="p">,</span> <span class="s2">&quot;orange&quot;</span><span class="p">,</span> <span class="s2">&quot;deepskyblue&quot;</span><span class="p">,</span> <span class="s2">&quot;darkblue&quot;</span><span class="p">,</span> <span class="s2">&quot;darkviolet&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">autocrop</span><span class="p">()</span>
    <span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">render_image</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">render_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_CellTypePrediction_32_0.png" src="../_images/notebooks_CellTypePrediction_32_0.png" />
</div>
</div>
<p>As a last step, let’s say you want to go back to a lower level. You can do this using the <code class="docutils literal notranslate"><span class="pre">la.set_label_level()</span></code> method as shown below.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="c1"># showing the fine-grained labels, which are the defaults</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">ds_with_subtypes</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">autocrop</span><span class="p">()</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">render_image</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">render_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="c1"># going back a level (so that T_tox cells are not split into memory and naive)</span>
<span class="n">_</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ds_with_subtypes</span><span class="o">.</span><span class="n">la</span><span class="o">.</span><span class="n">set_label_level</span><span class="p">(</span><span class="s2">&quot;labels_1&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">autocrop</span><span class="p">()</span>
    <span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">render_image</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">render_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_CellTypePrediction_34_0.png" src="../_images/notebooks_CellTypePrediction_34_0.png" />
</div>
</div>
</section>
<section id="Advanced-Cell-Type-Gating">
<h2>Advanced Cell Type Gating<a class="headerlink" href="#Advanced-Cell-Type-Gating" title="Link to this heading">¶</a></h2>
<p>The example above provided an overview on how to use marker positivity to gate different cell types. You can also encode more complex gating schemes with marker negativity, double-positivity, etc. Let’s look at some ways we can encode a more complex cell type hierarchy. Note that these are just for illustrative purposes and do not represent the most biologically sensible gating strategy. Furthermore, they are shown in <code class="docutils literal notranslate"><span class="pre">yaml</span></code> format for better readability.</p>
<ol class="arabic simple">
<li><p><strong>Double positivity:</strong> You can encode a cell type with double positivity. In the case below, we would only call a cell as naive cytotoxic, if both CD8 and CD45RA are positive.</p></li>
</ol>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>B:
  subtypes:
    - name: B_prol
      markers: [&quot;Ki-67+&quot;]
T:
  subtypes:
    - name: T_h
      markers: [&quot;CD4+&quot;]
    - name: T_tox
      markers: [&quot;CD8+&quot;, &quot;CD45RA+&quot;]
</pre></div>
</div>
<ol class="arabic simple" start="2">
<li><p><strong>Alternative markers:</strong> Alternatively, you could also define a cell type if either one of a set of markers is positive. Here, cells would be called T_tox if either CD45RO or CD45RA are positive.</p></li>
</ol>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>B:
  subtypes:
    - name: B_prol
      markers: [&quot;Ki-67+&quot;]
T:
  subtypes:
    - name: T_tox
      markers: [&quot;CD45RA+&quot;]
    - name: T_tox
      markers: [&quot;CD45RO+&quot;]
</pre></div>
</div>
<ol class="arabic simple" start="3">
<li><p><strong>Marker negativity:</strong> By using a minus sign after a marker, we call a cell if it is negative for that marker. For example, we could define T helper cells as being CD8 negative.</p></li>
</ol>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>B:
  subtypes:
    - name: B_prol
      markers: [&quot;Ki-67+&quot;]
T:
  subtypes:
    - name: T_h
      markers: [&quot;CD8-&quot;]
    - name: T_tox
      markers: [&quot;CD8+&quot;]
</pre></div>
</div>
<ol class="arabic simple" start="4">
<li><p><strong>Combined positivity and negativity:</strong> We can also use combinations of positive and negative markers to define certain cell subtypes.</p></li>
</ol>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>B:
  subtypes:
    - name: B_prol
      markers: [&quot;Ki-67+&quot;]
T:
  subtypes:
    - name: T_h
      markers: [&quot;CD4+&quot;, &quot;CD8-&quot;]
    - name: T_tox
      markers: [&quot;CD8+&quot;]
</pre></div>
</div>
</section>
</section>


              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="ImageProcessing.html" title="previous page">Image Processing</a>
    <a class='right-next' id="next-link" href="Neighborhoods.html" title="next page">Neighborhood Analysis</a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Matthias Meyer-Bender, Harald Vohringer<br/>
        
            &copy; Copyright 2024, Matthias Meyer-Bender, Harald Vohringer.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="../_static/js/index.d3f166471bb80abb5163.js"></script>


    
  </body>
</html>